pkg/operator/staticpod/controller/staticpodfallback/static_pod_fallback_condition.go

```diff
@@ -0,0 +1,86 @@
+package staticpodfallback
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/staticpod/startupmonitor/annotations"
+	operatorv1helpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+
+	"k8s.io/apimachinery/pkg/labels"
+	corev1listers "k8s.io/client-go/listers/core/v1"
+)
+
+// staticPodFallbackConditionController knows how to detect and report that a static pod was rolled back to a previous revision
+type staticPodFallbackConditionController struct {
+	operatorClient operatorv1helpers.OperatorClient
+
+	podLabelSelector labels.Selector
+	podLister        corev1listers.PodNamespaceLister
+
+	preconditionFulfilledFn func() bool
+}
+
+// New creates a controller that detects and report roll back of a static pod
+func New(targetNamespace string,
+	podLabelSelector labels.Selector,
+	operatorClient operatorv1helpers.OperatorClient,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	preconditionFulfilledFn func() bool,
+	eventRecorder events.Recorder) factory.Controller {
+	fd := &staticPodFallbackConditionController{
+		operatorClient:          operatorClient,
+		podLabelSelector:        podLabelSelector,
+		podLister:               kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().Pods().Lister().Pods(targetNamespace),
+		preconditionFulfilledFn: preconditionFulfilledFn,
+	}
+	return factory.New().WithSync(fd.sync).ResyncEvery(6*time.Minute).WithInformers(kubeInformersForNamespaces.InformersFor(targetNamespace).Core().V1().Pods().Informer()).ToController("StaticPodStateFallback", eventRecorder)
+}
+
+// sync sets/unsets a StaticPodFallbackRevisionDegraded condition if a pod that matches the given label selector is annotated with FallbackForRevision
+func (fd *staticPodFallbackConditionController) sync(_ context.Context, _ factory.SyncContext) (err error) {
+	degradedCondition := operatorv1.OperatorCondition{Type: "StaticPodFallbackRevisionDegraded", Status: operatorv1.ConditionFalse}
+	defer func() {
+		if _, _, updateError := operatorv1helpers.UpdateStatus(fd.operatorClient, operatorv1helpers.UpdateConditionFn(degradedCondition)); updateError != nil {
+			err = updateError
+		}
+	}()
```

Comment:
I think this will swallow the original error
https://play.golang.org/p/6fCSL5LaadB
maybe we can wrap it?

Reply:
worse: we unconditionally set the upper `degradedCondition` condition when some `return err` below is hit.

Reply:
thx